File: scripts/seed-suppliers.mjs
/**
 * Seed suppliers from the exported CSV into the database.
 * Run: node scripts/seed-suppliers.mjs
 */
import { createConnection } from "mysql2/promise";
import { readFileSync } from "fs";
import { parse } from "csv-parse/sync";
import * as dotenv from "dotenv";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const __dirname = dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: join(__dirname, "../.env") });

const DATABASE_URL = process.env.DATABASE_URL;
if (!DATABASE_URL) {
  console.error("DATABASE_URL not set");
  process.exit(1);
}

// Parse DATABASE_URL: mysql://user:pass@host:port/db
function parseMysqlUrl(url) {
  const u = new URL(url);
  return {
    host: u.hostname,
    port: parseInt(u.port || "3306"),
    user: u.username,
    password: u.password,
    database: u.pathname.slice(1),
    ssl: { rejectUnauthorized: false },
  };
}

// Clean HTML from description fields
function stripHtml(html) {
  if (!html) return "";
  return html
    .replace(/<[^>]*>/g, " ")
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&nbsp;/g, " ")
    .replace(/&quot;/g, '"')
    .replace(/\s+/g, " ")
    .trim();
}

// Determine credential stage based on whether credentials exist
function getCredentialStage(row) {
  const hasCredentials = row["username__13"]?.trim() || row["password"]?.trim();
  // Default: if credentials exist, require stage 2 to see them
  // Admin can change this per supplier
  return hasCredentials ? 2 : 2;
}

async function main() {
  // Read and parse CSV (UTF-16 encoded, tab-separated)
  const csvPath = join(__dirname, "../../upload/export_dir/export.csv");
  let content;
  try {
    content = readFileSync(csvPath, "utf-16le");
    // Remove BOM if present
    if (content.charCodeAt(0) === 0xFEFF) {
      content = content.slice(1);
    }
  } catch (e) {
    // Try the uploaded CSV directly
    const altPath = join(__dirname, "../../upload/pasted_file_b3itK8_export.csv");
    content = readFileSync(altPath, "utf-16le");
    if (content.charCodeAt(0) === 0xFEFF) {
      content = content.slice(1);
    }
  }

  const rows = parse(content, {
    delimiter: "\t",
    columns: true,
    skip_empty_lines: true,
    relax_quotes: true,
    relax_column_count: true,
  });

  console.log(`Parsed ${rows.length} rows from CSV`);

  const conn = await createConnection(parseMysqlUrl(DATABASE_URL));

  // Clear existing suppliers
  await conn.execute("DELETE FROM suppliers");
  console.log("Cleared existing suppliers");

  let inserted = 0;
  let errors = 0;

  for (let i = 0; i < rows.length; i++) {
    const row = rows[i];
    const name = row["listing_title"]?.trim();
    if (!name) continue;

    // Build image filename (will be updated with S3 URL later)
    const imageFilename = row["images"]?.trim() || null;

    // Clean categories and locations
    const categories = row["listing_category"]
      ?.split(";")
      .map((c) => c.trim())
      .filter(Boolean)
      .join(";") || null;

    const locations = row["locations_tags"]
      ?.split(";")
      .map((l) => l.trim())
      .filter(Boolean)
      .join(";") || null;

    // Clean website URLs (remove trailing commas)
    const cleanUrl = (url) => url?.trim().replace(/,\s*$/, "") || null;

    const supplier = {
      name,
      description: row["description"]?.trim() || null,
      shortDescription: row["short_description"]?.trim() || null,
      publicWebsite: cleanUrl(row["public_website"]),
      tradeWebsite: cleanUrl(row["trade_website"]) || cleanUrl(row["website_1"]) || cleanUrl(row["website"]),
      additionalWebsite: null,
      agencyId: row["agency_id"]?.trim() || null,
      loginUsername: row["username__13"]?.trim() || null,
      loginPassword: row["password"]?.trim() || null,
      commission: row["commission"]?.trim() || null,
      facebookUrl: cleanUrl(row["facebook"]),
      accountManager: row["account_manager"]?.trim() || null,
      phone: row["phone"]?.trim() || null,
      email: row["email"]?.trim() || null,
      generalNotes: row["general"]?.trim() || null,
      video1: row["video_1"]?.trim() || null,
      video2: row["video_2"]?.trim() || null,
      video3: row["video_3"]?.trim() || null,
      categories,
      locations,
      imageUrl: imageFilename ? `__PENDING__${imageFilename}` : null, // will be replaced with S3 URL
      adminUsername: row["admin_only_-_username"]?.trim() || null,
      adminPassword: row["admin_only_-_password"]?.trim() || null,
      adminNotes: row["admin_only_-_notes"]?.trim() || null,
      credentialStage: getCredentialStage(row),
      isActive: 1,
      sortOrder: i,
    };

    try {
      await conn.execute(
        `INSERT INTO suppliers 
          (name, description, shortDescription, publicWebsite, tradeWebsite, additionalWebsite,
           agencyId, loginUsername, loginPassword, commission, facebookUrl, accountManager,
           phone, email, generalNotes, video1, video2, video3, categories, locations, imageUrl,
           adminUsername, adminPassword, adminNotes, credentialStage, isActive, sortOrder)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        [
          supplier.name, supplier.description, supplier.shortDescription,
          supplier.publicWebsite, supplier.tradeWebsite, supplier.additionalWebsite,
          supplier.agencyId, supplier.loginUsername, supplier.loginPassword,
          supplier.commission, supplier.facebookUrl, supplier.accountManager,
          supplier.phone, supplier.email, supplier.generalNotes,
          supplier.video1, supplier.video2, supplier.video3,
          supplier.categories, supplier.locations, supplier.imageUrl,
          supplier.adminUsername, supplier.adminPassword, supplier.adminNotes,
          supplier.credentialStage, supplier.isActive, supplier.sortOrder,
        ]
      );
      inserted++;
    } catch (e) {
      console.error(`Error inserting "${name}":`, e.message);
      errors++;
    }
  }

  await conn.end();
  console.log(`\nDone! Inserted: ${inserted}, Errors: ${errors}`);
  console.log("Note: Image URLs are set to __PENDING__<filename> and will be updated after S3 upload.");
}

main().catch(console.error);

File: scripts/seed-suppliers.py
#!/usr/bin/env python3
"""
Seed suppliers from the exported CSV into the database.
Run: python3 scripts/seed-suppliers.py
"""
import csv
import io
import os
import sys
import re
import urllib.parse
import mysql.connector

# Load .env file
def load_env():
    env_path = os.path.join(os.path.dirname(__file__), "../.env")
    if os.path.exists(env_path):
        with open(env_path) as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith("#") and "=" in line:
                    key, _, val = line.partition("=")
                    os.environ.setdefault(key.strip(), val.strip().strip('"').strip("'"))

load_env()

DATABASE_URL = os.environ.get("DATABASE_URL", "")
if not DATABASE_URL:
    print("DATABASE_URL not set")
    sys.exit(1)

def parse_mysql_url(url):
    u = urllib.parse.urlparse(url)
    return {
        "host": u.hostname,
        "port": u.port or 3306,
        "user": u.username,
        "password": urllib.parse.unquote(u.password or ""),
        "database": u.path.lstrip("/"),
        "ssl_disabled": False,
    }

def clean_url(url):
    if not url:
        return None
    url = url.strip().rstrip(",").strip()
    return url if url else None

def get_credential_stage(row):
    has_creds = bool(row.get("username__13", "").strip() or row.get("password", "").strip())
    return 2  # Default: require stage 2 to see credentials

def main():
    # Read CSV
    csv_path = os.path.join(os.path.dirname(__file__), "../../upload/export_dir/export.csv")
    if not os.path.exists(csv_path):
        csv_path = os.path.join(os.path.dirname(__file__), "../../upload/pasted_file_b3itK8_export.csv")

    with open(csv_path, encoding="utf-16") as f:
        content = f.read()

    reader = csv.DictReader(io.StringIO(content), delimiter="\t")
    rows = list(reader)
    print(f"Parsed {len(rows)} rows from CSV")

    # Connect to DB
    cfg = parse_mysql_url(DATABASE_URL)
    conn = mysql.connector.connect(**cfg)
    cursor = conn.cursor()

    # Clear existing suppliers
    cursor.execute("DELETE FROM suppliers")
    conn.commit()
    print("Cleared existing suppliers")

    inserted = 0
    errors = 0

    for i, row in enumerate(rows):
        name = row.get("listing_title", "").strip()
        if not name:
            continue

        image_filename = row.get("images", "").strip() or None

        categories = ";".join(
            c.strip() for c in row.get("listing_category", "").split(";") if c.strip()
        ) or None

        locations = ";".join(
            l.strip() for l in row.get("locations_tags", "").split(";") if l.strip()
        ) or None

        trade_web = (
            clean_url(row.get("trade_website", "")) or
            clean_url(row.get("website_1", "")) or
            clean_url(row.get("website", ""))
        )

        image_url = f"__PENDING__{image_filename}" if image_filename else None

        values = (
            name,
            row.get("description", "").strip() or None,
            row.get("short_description", "").strip() or None,
            clean_url(row.get("public_website", "")),
            trade_web,
            None,  # additionalWebsite
            row.get("agency_id", "").strip() or None,
            row.get("username__13", "").strip() or None,
            row.get("password", "").strip() or None,
            row.get("commission", "").strip() or None,
            clean_url(row.get("facebook", "")),
            row.get("account_manager", "").strip() or None,
            row.get("phone", "").strip() or None,
            row.get("email", "").strip() or None,
            row.get("general", "").strip() or None,
            row.get("video_1", "").strip() or None,
            row.get("video_2", "").strip() or None,
            row.get("video_3", "").strip() or None,
            categories,
            locations,
            image_url,
            row.get("admin_only_-_username", "").strip() or None,
            row.get("admin_only_-_password", "").strip() or None,
            row.get("admin_only_-_notes", "").strip() or None,
            get_credential_stage(row),
            1,  # isActive
            i,  # sortOrder
        )

        try:
            cursor.execute(
                """INSERT INTO suppliers
                    (name, description, shortDescription, publicWebsite, tradeWebsite, additionalWebsite,
                     agencyId, loginUsername, loginPassword, commission, facebookUrl, accountManager,
                     phone, email, generalNotes, video1, video2, video3, categories, locations, imageUrl,
                     adminUsername, adminPassword, adminNotes, credentialStage, isActive, sortOrder)
                   VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)""",
                values,
            )
            inserted += 1
        except Exception as e:
            print(f"Error inserting '{name}': {e}")
            errors += 1

        if (inserted + errors) % 50 == 0:
            conn.commit()
            print(f"  Progress: {inserted} inserted, {errors} errors...")

    conn.commit()
    cursor.close()
    conn.close()

    print(f"\nDone! Inserted: {inserted}, Errors: {errors}")
    print("Note: Image URLs are set to __PENDING__<filename> — run the image upload script next.")

if __name__ == "__main__":
    main()
